pkg/controller/azvolumeattachment.go

```diff
@@ -0,0 +1,546 @@
+	var azVolumeAttachment v1alpha1.AzVolumeAttachment
+	if err := r.client.Get(ctx, request.NamespacedName, &azVolumeAttachment); err != nil {
+		// if the azVolumeAttachment object can no longer be found, it's been deleted so do not requeue
+		if errors.IsNotFound(err) {
+			return reconcile.Result{}, nil
+		}
+		// if the GET failure is triggered by other errors, log it and requeue the request
+		klog.Errorf("failed to fetch azvolumeattachment object with namespaced name %s: (%v)", request.NamespacedName, err)
+		return reconcile.Result{Requeue: true}, err
+	}
+
+	// this is a creation event
+	if azVolumeAttachment.Status == nil {
+		// First, attach the volume to the specified node
+		if err := r.AttachVolume(ctx, azVolumeAttachment.Spec.UnderlyingVolume, azVolumeAttachment.Spec.NodeName); err != nil {
+			klog.Errorf("failed to attach volume %s to node %s: (%v)", azVolumeAttachment.Spec.UnderlyingVolume, azVolumeAttachment.Spec.NodeName, err)
+			return reconcile.Result{Requeue: true}, err
+		}
+
+		// If it is the first time seeing this underlying volume, sync scMap and volMap
+		if _, ok := r.volMap[azVolumeAttachment.Spec.UnderlyingVolume]; !ok {
+			if err := r.SyncStorageClass(ctx); err != nil {
+				return reconcile.Result{Requeue: true}, err
+			}
+		}
+
+		// update the status
+		updated := azVolumeAttachment.DeepCopy()
+		updated.Status = &v1alpha1.AzVolumeAttachmentStatus{}
+		updated.Status.AttachmentTier = updated.Spec.AttachmentTier
+		updated.Status.AttachmentState = v1alpha1.Attached
+
+		if azVolumeAttachment.Spec.AttachmentTier == v1alpha1.Primary {
+			primary, primaryOk := r.primaryMap[azVolumeAttachment.Spec.UnderlyingVolume]
+			// this is a case when previous primary attachment object had not been deleted for some reason,
+			// in this case, mark the old primary detach required, and update the primary map
+			if primaryOk && primary.Name != azVolumeAttachment.Name {
+				if err := r.TriggerDetach(ctx, primary); err != nil {
+					// if detach failed, requeue
+					return reconcile.Result{Requeue: true}, err
+				}
+				// this is a case when the attachment object is created for the first time for the specified underlying volume
+				// in this case, create n replicas
+			} else {
+				if err := r.ManageReplicaForVolume(ctx, azVolumeAttachment.Spec.UnderlyingVolume); err != nil {
+					return reconcile.Result{Requeue: true}, err
+				}
+			}
+		}
+
+		if _, err := r.azVolumeClient.DiskV1alpha1().AzVolumeAttachments(r.namespace).UpdateStatus(ctx, updated, metav1.UpdateOptions{}); err != nil {
+			klog.Errorf("failed to update status for azvolumeattachment object %s: (%v)", azVolumeAttachment.Name, err)
+			return reconcile.Result{Requeue: true}, err
+		}
+
+		// if status update was successful add the object to either the replica or primary map
+		if azVolumeAttachment.Spec.AttachmentTier == v1alpha1.Primary {
+			r.primaryMap[azVolumeAttachment.Spec.UnderlyingVolume] = azVolumeAttachment
```

Comment:
Since the reconciler concurrency is > 1, you will need to synchronize access to any maps updated by it or in parallel to it.

Reply:
Good point. I will give it some more thoughts.